### DISCUSSION TITLE
Update Conda Environment to Mamba

### DISCUSSION THREAD

**Discussion by TLWise** [idaholab/moose#25485] [Q&A Getting Started]
On one of my computers I have a version of MOOSE that is Conda based. I would like to upgrade to the newest version without deleting my current research efforts. What is the most efficient way to perform such an update?

**Comment by GiudGiud** [comment]
Hello

You ll want to rebase your work on top of the current version of MOOSE
```
git fetch origin
git rebase origin/master
```
then on each commit you may have to fix up conflicts with the current version of MOOSE if you modified things that got modified between the previous branch you used and now

Guillaume

**Reply by TLWise** [reply]
Here is what came up from the fetch origin:
From https://github.com/idaholab/moose
   247d605b0d..6ad394d0e8  master                -> origin/master
 * [new branch]            2020-01-24-release    -> origin/2020-01-24-release
 * [new branch]            2021-05-18-release    -> origin/2021-05-18-release
 * [new branch]            2021-09-15-release    -> origin/2021-09-15-release
 * [new branch]            2022-06-10-release    -> origin/2022-06-10-release
   c7ac971e03..49df46a08c  devel                 -> origin/devel
 * [new branch]            june_release          -> origin/june_release
   73dd22b5a5..49df46a08c  next                  -> origin/next
 * [new branch]            scheduled             -> origin/scheduled
 * [new branch]            sqa-integration-19015 -> origin/sqa-integration-19015
 * [new tag]               2021-09-15            -> 2021-09-15
 * [new tag]               2022-06-10            -> 2022-06-10
 * [new tag]               2023-06-13            -> 2023-06-13

**Reply by TLWise** [reply]
git rebase origin/master
error: cannot rebase: You have unstaged changes.
error: Please commit or stash them.

**Reply by TLWise** [reply]
Current branch master is up to date.

**Reply by TLWise** [reply]
Now that I am up to date with the current version. What is the best way to proceed to update to the mamba environment?

**Reply by GiudGiud** [reply]
mamba update —all

**Reply by GiudGiud** [reply]
For the previous message you got you ll have to commit (or stash) all your changes

**Reply by TLWise** [reply]
After performing the update and using the command "make -j4", I am getting errors such as this:

/home/twise/projects/moose/framework/build/header_symlinks/MooseFunctor.h:878:24: error: no type named 'GradientType' in 'class Moose::FunctorBase<MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<100> >, true> >'
  878 |   virtual GradientType evaluateGradient(const ElemArg & elem, const StateArg & state) const override
      |                        ^~~~~~~~~~~~~~~~

**Reply by GiudGiud** [reply]
ok so commit / save everything you have in that moose **folder** (that part is important, every single file that is not committed or even not tracked will be lost if you dont save it)

then go in moose/
```
git clean -xfd
```
then try to rebuild

**Reply by TLWise** [reply]
when I attempt to commit, I get the following prompt:

Username for 'https://github.com': wisetl@email.sc.edu
Password for 'https://wisetl@email.sc.edu@github.com': password

remote: Support for password authentication was removed on August 13, 2021.
remote: Please see https://docs.github.com/en/get-started/getting-started-with-git/about-remote-repositories#cloning-with-https-urls for information on currently recommended modes of authentication.
fatal: Authentication failed for 'https://github.com/idaholab/moose.git/'

My original username and password no longer function for me to upload to the git repository. I created a token today, but I am not sure of how to use it.

**Reply by GiudGiud** [reply]
you should use SSH keys
Please follow these instructions to add the SSH keys both on your machine (see the pre-requisites on that page) and on github
https://docs.github.com/en/authentication/connecting-to-github-with-ssh/adding-a-new-ssh-key-to-your-github-account

**Reply by TLWise** [reply]
For the key setup, is there a way to get around this error?:

@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@         WARNING: UNPROTECTED PRIVATE KEY FILE!          @
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
Permissions 0644 for '/home/twise/.ssh/xxxx.pub' are too open.
It is required that your private key files are NOT accessible by others.
This private key will be ignored.

**Reply by GiudGiud** [reply]
yes use chmod to change the permissions of the file
chmod 700 /home/twise/.ssh/xxxx.pub

**Comment by TLWise** [comment]
I performed the following:
git fetch origin
git rebase origin/master

I got the following:
First, rewinding head to replay your work on top of it...
error: Updating the following directories would lose untracked files in them:
	modules/heat_conduction/src

Aborting
fatal: Could not detach HEAD

**Reply by milljm** [reply]
Untracked files are files you've created, that you have not yet told git to track. If you don't care about these files, you can clean them up with `git clean` commands... However, I am betting that is not what you want to do.


Looking over this thread, I want to spell out a couple of phrases, in an attempt to drive a couple points home. Git is complex, no argument there! I make mistakes all the time. And some of `git`s naming conventions don't help. Also, I believe you are attempting to rebase from the wrong remote. `origin` should be `upstream` if you are applying updates _and_ if you also have your own github account hosting a fork of moose (which I believe you do).

Common workflow practice when making and saving new changes:
```pre
1. You need to tell git to track these files, and save them.
                           ^^^^^ `git add`        ^^^^  `git commit`

2. Optional; once you have saved your work you can also
             save your work to your github account
             ^^^^ `git push origin`

3. Finally you can apply updates from us
                   ^^^^^^^^^^^^^ `git fetch upstream; git rebase upstream/master`

Repeat.
```
